### DISCUSSION TITLE
Custom uploader without url

### DISCUSSION THREAD

**Discussion by bruceharrison1984** [rpldy/react-uploady#278] [Q&A]
**Is your feature request related to a problem? Please describe.**
I'm wanting to use uploady for images, but I am using Supabase as my backend. It doesn't support SignedUrls for uploads, so the only way it can be used is via the SDK. I'm sure there are other scenarios where this happens as well.

**Describe the solution you'd like**
A way to pass a custom upload function that allows me to send the file via SDK code.

**Describe alternatives you've considered**
Since Supabase doesn't allow generation of signed urls for upload, there aren't many outside of creating a proxy endpoint to receive the file FormData, and pass it along from there.

**Additional context**
None, but great work on this!

**Comment by yoavniran** [comment]
Hi @bruceharrison1984, I believe what you're asking for is already possible with Uploady as it is today.
Its pretty much what the custom senders ([chunked](https://github.com/rpldy/react-uploady/tree/master/packages/core/chunked-sender), [tus](https://github.com/rpldy/react-uploady/tree/master/packages/core/tus-sender), [mock](https://github.com/rpldy/react-uploady/tree/master/packages/core/mock-sender)) of the library do. They replace the default ([xhr](https://github.com/rpldy/react-uploady/tree/master/packages/core/sender)) sender with their own send method.

See [here](https://github.com/rpldy/react-uploady/blob/d7b232c38297320714ced62497300270a17d9201/packages/core/sender/types/index.d.ts#L31) the signature of the send method. You can pass your custom send method as a prop to Uploady.

Im not familiar with Supabase SDK but I expect it shouldnt be difficult to integrate since you will be getting the File object from the Uploader in your custom send method.

I'll move this thread to the discussions section.

**Comment by bruceharrison1984** [comment]
That you very much @yoavniran! I read through the docs but didn't see any explicit mention of this so I incorrectly presumed it wasn't supported. I'll take a look, and thank you!

**Reply by yoavniran** [reply]
no problem. feel free to update me how the integration goes.

**Reply by bruceharrison1984** [reply]
My super naive implementation is as follows:

```js
import { BatchItem, FILE_STATES } from '@rpldy/shared';
import { OnProgress, SendMethod, SendOptions, SendResult } from '@rpldy/sender';
import { STORAGE_BUCKET } from '@/utils/constants';
import { SupabaseClient } from '@supabase/supabase-js';

export const SupabaseSender = (
  supabase: SupabaseClient,
  storagePath: string
) => {
  const process: SendMethod = (
    items: BatchItem[],
    url: string | undefined = '',
    options: SendOptions,
    onProgress?: OnProgress
  ) => {
    let result: SendResult;

    supabase.storage
      .from(STORAGE_BUCKET)
      .upload(storagePath, <Blob>(<unknown>items[0].file));

    return {
      request: new Promise((resolve, reject) =>
        resolve({ status: 1, state: FILE_STATES.FINISHED, response: null })
      ),
      abort: () => false,
      senderType: '?',
    };
  };

  return process;
};
```

Lots of room for improvement and error handling, but this gets the file in to Supabase as expected. 

Thanks for pointing me in the right direction.

**Reply by yoavniran** [reply]
very cool. I'd definitely add error handling and waiting  for the  the upload to finish before returning  control back  to Uploady.

**Reply by bruceharrison1984** [reply]
I plan on cleaning this up quite a bit (and actually using the request prop). My question to you would be once finished, would you be interested in me submitting a PR to add Supabase as a provider to this repo, or should I just publish it as a separate package?

**Reply by yoavniran** [reply]
I think a separate repo will be ideal. Similar to this [drive-uploady](https://github.com/rpldy/drive-uploady) repo.
If you'd like I can open a repo for it under [rpldy org](https://github.com/rpldy) and you can be a maintainer.

**Reply by thorwebdev** [reply]
@bruceharrison1984 thanks so much for sharing this! I was wondering, did you ever end up releasing this custom sender? Would love to help amplify. 

On a related note, we're looking into supporting tus as an uploader for Supabase Storage soon, so this could be another path in the future.

**Reply by thorwebdev** [reply]
@yoavniran would you mind creating the `supabase-uploady` repo and adding myself as a maintainer? I'd love to contribute :)

**Reply by thorwebdev** [reply]
This is the supabase sender I've come up with:

```js
import { BatchItem, FILE_STATES } from "@rpldy/shared";
import { SendMethod, SendResult } from "@rpldy/sender";
import { SupabaseClient } from "@supabase/supabase-js";

const validateItems = (items: BatchItem[]) => {
  if (items.length > 1) {
    throw new Error(
      "Uploady Supabase Storage Sender - only 1 file can be uploaded per request (use concurrent = 1)"
    );
  }

  if (!items[0].file) {
    throw new Error(
      "Uploady Supabase Storage Sender - uploaded item must be file"
    );
  }
};

export const supabaseStorageSender = ({
  supabase,
  fileName,
  fileNamePrefix,
  bucket,
}: {
  supabase: SupabaseClient;
  fileName?: string;
  fileNamePrefix?: string;
  bucket: string;
}) => {
  const process: SendMethod = (items: BatchItem[]) => {
    let result: SendResult = {
      request: new Promise((resolve, reject) => {
        try {
          validateItems(items);
          supabase.storage
            .from(bucket)
            .upload(
              fileName ??
                `${fileNamePrefix ? fileNamePrefix + "_" : ""}${
                  items[0].file.name
                }`,
              <Blob>(<unknown>items[0].file)
            )
            .then(({ data, error }) => {
              console.log({ data, error });
              if (data)
                resolve({
                  status: 200,
                  state: FILE_STATES.FINISHED,
                  response: {
                    ...data,
                    publicUrl: supabase.storage
                      .from(bucket)
                      .getPublicUrl(data.path).data.publicUrl,
                  },
                });
              if (error) reject(error);
            });
        } catch (error) {
          reject(error);
        }
      }),
      abort: () => false,
      senderType: "supabase-storage-sender",
    };

    return result;
  };

  return process;
};
```

And then you can use it as follows:

```js
const UploadButtonWithEvents = () => {
    useItemFinishListener((file) => {
      setPhotoName(file.uploadResponse.path);
      setOriginalPhoto(file.uploadResponse.publicUrl);
    });

    return <UploadButton />;
  };

  const UploadDropZone = ({ supabase }: { supabase: SupabaseClient }) => (
    <Uploady
      multiple={false}
      send={supabaseStorageSender({
        supabase,
        bucket: "rooms",
        fileNamePrefix: session?.user.name ?? "",
      })}
    >
      <UploadButtonWithEvents />
      <UploadPreview />
    </Uploady>
```

Would love to hear your feedback :)

**Reply by yoavniran** [reply]
great start @thorwebdev 
I've created the [repo](https://github.com/rpldy/supabase-uploady) and added you.

Note that there's quite some work to do in order for this code to be ready to be published as a _supabase-uploady_ provider. 

In needs to follow the same pattern found in drive-uploady:

- a [wrapper around Uploady](https://github.com/rpldy/drive-uploady/blob/master/src/DriveUploady.js) 
- an [Uploader Enhancer ](https://github.com/rpldy/drive-uploady/blob/master/src/getDriveEnhancer.js)
- Tests (unit + integration) will be best
- Readme
- I dont see implementation of progress. Is it not possible to get progress for the upload from the SupabaseClient?
- No implementation of abort. Is it not possible?

Let me know what I can help with

**Reply by thorwebdev** [reply]
Thanks, @yoavniran.

We're currently working on supporting tus, so I'm wondering how worthwhile it is to built a separate wrapper. Let me test the tus-uploady first and see how that goes.

**Reply by yoavniran** [reply]
@thorwebdev plz let me know if I should keep the [supabase repo](https://github.com/rpldy/supabase-uploady) open or should I delete?
Thanks

**Reply by thorwebdev** [reply]
Since we now have [TUS support](https://supabase.com/docs/guides/storage/uploads#resumable-upload) I believe we can close out the supabase specific repo. Thanks and sorry for the hassle!
